src/awkward/_connect/numba/layoutbuilder.py

```diff
@@ -0,0 +1,2547 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+
+import math
+
+import numba
+import numba.core.typing.npydecl
+import numpy as np
+
+import awkward as ak
+from awkward._connect.numba.growablebuffer import GrowableBuffer, GrowableBufferType
+from awkward._nplikes.numpylike import ArrayLike
+from awkward._typing import final
+
+
+class LayoutBuilder:
+    def __init__(self, builder, *, parameters=None):
+        self._builder = builder
+        self._parameters = parameters
+
+    @property
+    def builder(self):
+        return self._builder
+
+    def clear(self):
+        return self._builder.clear()
+
+    def is_valid(self, error: str):
+        return self._builder.is_valid(error)
+
+    def _type(self, typestrs):
+        raise NotImplementedError
```

Comment:
This looks like an abstract base class for all of the Python-accessible builders, a role similar to [ak.forms.Form](https://github.com/scikit-hep/awkward/blob/f2783ba480b5f9f668b687e327d0a5a453b82e5b/src/awkward/forms/form.py#L335). It doesn't look like an outer wrapper in the style of [ak.types.ArrayType](https://github.com/scikit-hep/awkward/blob/f2783ba480b5f9f668b687e327d0a5a453b82e5b/src/awkward/types/arraytype.py#L11) or [ArrayBuilder](https://github.com/scikit-hep/awkward/blob/main/awkward-cpp/include/awkward/builder/ArrayBuilder.h), and layout builders shouldn't have a wrapper, anyway (because they're targeting layouts, not high-level arrays).

However,

  * An abstract base class shouldn't be directly constructable, and this one has an `__init__` that looks, to a user, like something they should use. We didn't prohibit direct construction of `Form` by putting an exception in `Form.__init__` (although we could have, to make it more clear), but in the place where the base class has an init-helper, its name is [_init](https://github.com/scikit-hep/awkward/blob/f2783ba480b5f9f668b687e327d0a5a453b82e5b/src/awkward/forms/form.py#L345), to make it more clear that this isn't something users should use.
  * This `__init__` takes a `builder` argument and exposes a `builder` parameter, but not all subclasses of `LayoutBuilder` have a single child, which is anyway named `content`, rather than `builder`. The abstract base class shouldn't define a `builder`.
  * `clear` and `is_valid` seem to be methods used on all subclasses, so it's good to have them here with `raise NotImplementedError` (or better, `raise AssertionError("missing implementation")`) to make sure that none of them go unimplemented. This can be done better with Python-typing "protocols," though this PR should not get involved in adding Python type hints; that's too open-ended of a project.
  * All `LayoutBuilders` have lengths, so `__len__` would be another good method to provide a stub for here.
  * Also, all `LayoutBuilders` would have a `parameters` property. Since that's generic (there's no subclass-specific `parameters` handling), this class can handle that entirely, and the `parameters` property can be removed from subclasses.
  * I don't know why there's a `_type` method, or why it takes `typestrs`. On the other hand, `LayoutBuilders` will all need to produce Forms, so stubbing a `form` property at this level would be a good idea.

```suggestion
class LayoutBuilder:
    def _init(self, parameters):
        self._parameters = parameters

    @property
    def parameters(self):
        return self._parameters

    def __len__(self):
        raise AssertionError("missing implementation")

    @property
    def form(self):
        raise AssertionError("missing implementation")

    def clear(self):
        raise AssertionError("missing implementation")

    def is_valid(self, error: str):
        raise AssertionError("missing implementation")
```

Reply:
The `parameters` won't be all that useful in a Numba-compiled function, since it returns a JSON-like dict of data. Its data type would differ from one layout builder's `parameters` to another's, but a single builder's `parameters` would at least be a compile-time constant.

It's pretty common for a single parameter to be a string, and perhaps this is the only case that should be implemented in the compiled context. To do that, we'll need a function that only returns one parameter (or `None`). At the Python level, that would be a method on `LayoutBuilder` like

```python
    def parameter(self, name):
        return self._parameters.get(name)
```

I'll show the Numba-typing for this below.

Reply:
It also looks like you implemented a `length` property, rather than `__len__`, in the compiled context, so let's change `__len__` to

```python
    @property
    def length(self):
        raise AssertionError("missing implementation")
```